execution/block-partitioner/src/pre_partition/connected_component/mod.rs

```diff
@@ -0,0 +1,161 @@
+        // The workaround is to only fix the group set and their sizes for now, then schedule, and materialize the txn groups at the very end (when assigning groups to shards).
+        let group_metadata: Vec<(usize, usize)> = txns_by_set
+            .iter()
+            .enumerate()
+            .flat_map(|(set_idx, txns)| {
+                let num_big_chunks = txns.len() / group_size_limit;
+                let reminder_chunk_size = txns.len() % group_size_limit;
+                let mut ret = vec![(set_idx, group_size_limit); num_big_chunks];
+                if reminder_chunk_size >= 1 {
+                    ret.push((set_idx, reminder_chunk_size));
+                }
+                ret
+            })
+            .collect();
+
+        // Assign groups to shards using longest-processing-time first scheduling.
+        let tasks: Vec<u64> = group_metadata
+            .iter()
+            .map(|(_, size)| (*size) as u64)
+            .collect();
+        let (_longest_pole, shards_by_group) =
+            longest_processing_time_first(&tasks, state.num_executor_shards);
+
+        // Prepare `groups_by_shard`: a mapping from a shard to the txn groups assigned to it.
+        let mut groups_by_shard: Vec<Vec<usize>> = vec![vec![]; state.num_executor_shards];
+        for (group_id, shard_id) in shards_by_group.into_iter().enumerate() {
+            groups_by_shard[shard_id].push(group_id);
+        }
+
+        let mut pre_partitioned_idx0s: Vec<Vec<TxnIdx0>> = vec![vec![]; state.num_executor_shards];
+        for (shard_id, group_ids) in groups_by_shard.into_iter().enumerate() {
+            for group_id in group_ids.into_iter() {
+                let (set_id, amount) = group_metadata[group_id];
+                for _ in 0..amount {
+                    let txn_idx0 = txns_by_set[set_id].pop_front().unwrap();
+                    pre_partitioned_idx0s[shard_id].push(txn_idx0);
+                }
+            }
+        }
+
+        // Prepare `state.idx1_to_idx0` and `state.start_txn_idxs_by_shard`.
+        let mut i1 = 0;
+        for (shard_id, txn_idxs) in pre_partitioned_idx0s.iter().enumerate() {
+            state.start_txn_idxs_by_shard[shard_id] = i1;
+            for &i0 in txn_idxs {
+                state.idx1_to_idx0[i1] = i0;
+                i1 += 1;
+            }
+        }
+
+        // Prepare `state.pre_partitioned`.
+        state.pre_partitioned = (0..state.num_executor_shards)
+            .map(|shard_id| {
+                let start = state.start_txn_idxs_by_shard[shard_id];
+                let end: TxnIdx1 = if shard_id == state.num_executor_shards - 1 {
+                    state.num_txns()
+                } else {
+                    state.start_txn_idxs_by_shard[shard_id + 1]
+                };
+                (start..end).collect()
+            })
+            .collect();
+
+        state.thread_pool.spawn(move || {
```

Comment:
Is this needed ? That is, does dropping objects in another thread gives better perf ?
Intuitively, I would assume that compilers would generally optimize deleting resources in the same thread as they go out of scope. Calling drop in other threads might prevent such compiler optimizations.

I am not a rust expert, but would love to learn more about dropping resources in other threads.

Reply:
Not an expert either, but from my observation, dropping large objects is slow, and this trick indeed helps reduce latency.

BlockSTM code also use the same trick.
https://github.com/runtian-zhou/libra/blob/0c943ae7639393fd492a0a6660ddb4f026bff5ae/diem-move/parallel-executor/src/executor.rs#L226

Reply:
Hmm ok. I found this online https://stackoverflow.com/questions/73709474/how-bad-is-it-to-move-the-drop-of-a-large-struct-to-another-thread. It maybe good to drop in a separate thread.

I will ask this as a general question in rust channel, lets see if somebody as has any better insights

Reply:
maybe spawn_blocking is better than spawn, but lets leave this for now. 
We need to benchmark to be sure, which we can do independently some other time.
(https://aptos-org.slack.com/archives/C03A9SXHGJC/p1693963215339909)